### PR TITLE
Allow symbol data to be substituted at the time of applying the Operator

### PR DESCRIPTION
**Pull request by navjotk**
In response to #220 
This PR adds:
- The ability to do `op.apply(m=m1)`
- A test for this

There is one tiny caveat in this implementation though. In order to do `op.apply(m=m1)`, both `m` and `m1` need to have the same `name` property. In case this is not true, the expected syntax is `op.apply(m=m1.data)`. 
Edit: Never mind, the above has been fixed as well. `op.apply(m=m1)` would work regardless of the symbol names.

### CODE REVIEW

File: tests/test_stencil_arithmetic.py [hunk 1]
```diff
@@ -140,3 +140,24 @@ def test_arithmetic_indexed_open_loops(i, j, k, l, expr, result):
     eqn = eval(expr)
     StencilKernel(eqn)(fa)
     assert np.allclose(fa.data[1, 1:-1, 1:-1], result[1:-1, 1:-1], rtol=1e-12)
+
+
+def test_override(i, j, k, l):
+    """Test that the call-time overriding of Operator arguments works"""
+    a = symbol(name='a', dimensions=(i, j, k, l), value=2., mode='indexed').base.function
+    a1 = symbol(name='a', dimensions=(i, j, k, l), value=3., mode='indexed').base.function
```
Comment (mlange05) [inline tests/test_stencil_arithmetic.py]: I wonder why these two symbols not cache on the SymPy symbol-level? The test seems to work though, so I'm happy to go ahead.

File: tests/test_stencil_arithmetic.py [hunk 2]
```diff
@@ -140,3 +140,24 @@ def test_arithmetic_indexed_open_loops(i, j, k, l, expr, result):
...
+    op(a=a2)
+    shape = [d.size for d in [i, j, k, l]]
+
+    assert(np.allclose(a.data, np.zeros(shape) + 5))
+    assert(np.allclose(a1.data, np.zeros(shape) + 6))
+    assert(np.allclose(a2.data, np.zeros(shape) + 7))
+
+
+if __name__ == "__main__":
+    test_override(i(), j(), k(), l())
```
Comment (mlange05) [inline tests/test_stencil_arithmetic.py]: I don't like hijacking the main for one particular test. Why don't you use `py.test -k` to run individual tests instead?

File: devito/stencilkernel.py [hunk 3]
```diff
@@ -127,10 +128,22 @@ def arguments(self, *args, **kwargs):
         arguments = OrderedDict([(arg.name, arg) for arg in self.parameters])
         dim_sizes = {}
 
+        # Have we been provided substitutes for symbol data?
+        # Only SymbolicData can be overridden with this route
+        r_args = [f_n for f_n, f in arguments.items() if isinstance(f, SymbolicData)]
```
Comment (FabioLuporini) [inline devito/stencilkernel.py]: ``if f.is_SymbolicData`` ?

File: tests/test_stencil_arithmetic.py (hunk 1 above)
Comment (mloubout) [inline tests/test_stencil_arithmetic.py]: >I wonder why these two symbols not cache on the SymPy symbol-level? 
Does this mean some symbols may never clear their data, would be problematic

**Comment by mlange05** [comment]
Ok, so the cache misses on new `SymbolicData` instantiations with the same name are actually a known problem that was flagged in PR #164. There is an xfailing test already in `test_symbol_caching.py` that highlights this. I think filing a separate issue for this should be ok, so we can go ahead and land this PR.

File: tests/test_stencil_arithmetic.py (hunk 2 above)
Comment (mlange05) [inline tests/test_stencil_arithmetic.py]: Ping

File: devito/stencilkernel.py (hunk 3 above)
Comment (mlange05) [inline devito/stencilkernel.py]: Ping

Comment (navjotk) [inline devito/stencilkernel.py]: The objects here that are not `SymbolicData` are `BufferedDimension` objects. These objects don't have this property. This leads to `AttributeError: 'BufferedDimension' object has no attribute 'is_SymbolicData'`

**Comment by mlange05** [comment]
Looks good to me. Will merge once travis is happy after update.

**Pull request merged**